### PR TITLE
[FIXED JENKINS-28857] Catch the case in which LDAP times out after some seconds

### PR DESCRIPTION
**Pull request by fbelzunc**
`com.sun.jndi.ldap.Connection.readReply()` apparently times out after 15secs with an error like this:

```
Caused by: javax.naming.NamingException: LDAP response read timed out, timeout used:-1ms.; remaining name 'DC=example,DC=com'
    at com.sun.jndi.ldap.Connection.readReply(Unknown Source)
    at com.sun.jndi.ldap.LdapClient.getSearchReply(Unknown Source)
    at com.sun.jndi.ldap.LdapClient.search(Unknown Source)
    at com.sun.jndi.ldap.LdapCtx.doSearch(Unknown Source)
    at com.sun.jndi.ldap.LdapCtx.searchAux(Unknown Source)
    at com.sun.jndi.ldap.LdapCtx.c_search(Unknown Source)
    at com.sun.jndi.ldap.LdapCtx.c_search(Unknown Source)
    at com.sun.jndi.toolkit.ctx.ComponentDirContext.p_search(Unknown Source)
    at com.sun.jndi.toolkit.ctx.PartialCompositeDirContext.search(Unknown Source)
    at com.sun.jndi.toolkit.ctx.PartialCompositeDirContext.search(Unknown Source)
    at hudson.plugins.active_directory.LDAPSearchBuilder.search(LDAPSearchBuilder.java:97)
    at hudson.plugins.active_directory.ActiveDirectoryUnixAuthenticationProvider.chainGroupLookup(ActiveDirectoryUnixAuthenticationProvider.java:469)
    at hudson.plugins.active_directory.ActiveDirectoryUnixAuthenticationProvider.resolveGroups(ActiveDirectoryUnixAuthenticationProvider.java:453)
    at hudson.plugins.active_directory.ActiveDirectoryUnixAuthenticationProvider.retrieveUser(ActiveDirectoryUnixAuthenticationProvider.java:318)
    ... 44 more
```
Note that the source code (or that of JDK7) doesn't explain why one would ever get this error message. if readTimeout<0, then waited shall never be true.

This is reported as JDK-8062947.

https://issues.jenkins-ci.org/browse/JENKINS-28857

@reviewbybees

### CODE REVIEW

**Comment by jtnord** [comment]
sorry - commented on the commit not the PR :-(

no :bugs: but I feel my :bee: is missing a wing.

**Comment by fbelzunc** [comment]
@jtnord Would you mind to review again now that I have addressed your comments, please?

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java [hunk 1]
```diff
@@ -436,9 +436,19 @@ private String getPrincipalName(String username, String domainName) {
...
-                    long duration = (System.currentTimeMillis() - start) / TimeUnit2.SECONDS.toMillis(1);
+                    long start = System.nanoTime();
+                    boolean found = false;
+                    long duration;
+                    duration = TimeUnit2.NANOSECONDS.toSeconds(System.nanoTime() - start);
+                    try {
+                        found = chainGroupLookup(domainDN, userDN, context, groups);
+                    } catch (NamingException e) {
+                        if (e.getMessage().contains("LDAP response read timed out")) {
+                            LOGGER.log(Level.WARNING, String.format("LDAP response read time out. AD will fall back to recursive lookup", e);
```
Comment (jtnord) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java]: missing closing ')'

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java [hunk 2]
```diff
@@ -436,9 +436,19 @@ private String getPrincipalName(String username, String domainName) {
                 switch (groupLookupStrategy) {
                 case AUTO:
                     // try the accurate one first, and if it's too slow fall back to recursive in the hope that it's faster
-                    long start = System.currentTimeMillis();
-                    boolean found = chainGroupLookup(domainDN, userDN, context, groups);
-                    long duration = (System.currentTimeMillis() - start) / TimeUnit2.SECONDS.toMillis(1);
+                    long start = System.nanoTime();
+                    boolean found = false;
+                    long duration;
+                    duration = TimeUnit2.NANOSECONDS.toSeconds(System.nanoTime() - start);
```
Comment (jtnord) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java]: duration will be exceptionally small here - as you have not called the LDAP server yet!

**Comment by jtnord** [comment]
:bug: duration is calculated before actually calling the method that may take a while.

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java (hunk 2 above)
Comment (fbelzunc) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java]: Ups!

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java [hunk 3]
```diff
@@ -436,9 +436,19 @@ private String getPrincipalName(String username, String domainName) {
...
-                    long duration = (System.currentTimeMillis() - start) / TimeUnit2.SECONDS.toMillis(1);
+                    long start = System.nanoTime();
+                    boolean found = false;
+                    long duration = 0;
+                    try {
+                        found = chainGroupLookup(domainDN, userDN, context, groups);
+                        duration = TimeUnit2.NANOSECONDS.toSeconds(System.nanoTime() - start);
+                    } catch (NamingException e) {
+                        if (e.getMessage().contains("LDAP response read timed out")) {
+                            LOGGER.log(Level.WARNING, String.format("LDAP response read time out. AD will fall back to recursive lookup", e));
```
Comment (jtnord) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java]: `String.format("LDAP response read time out. AD will fall back to recursive lookup", e)`.equals("LDAP response read time out. AD will fall back to recursive lookup")

**Comment by rsandell** [comment]
:bee:

**Comment by jtnord** [comment]
:bee: but there is another issue that probably deserves a followup in [line 452](https://github.com/jenkinsci/active-directory-plugin/pull/20/files#diff-14402b0dc4dd755ecdc24a8b4dee5188R452)

File: src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java [hunk 4]
```diff
@@ -436,9 +436,19 @@ private String getPrincipalName(String username, String domainName) {
...
+                        found = chainGroupLookup(domainDN, userDN, context, groups);
+                        duration = TimeUnit2.NANOSECONDS.toSeconds(System.nanoTime() - start);
+                    } catch (NamingException e) {
+                        if (e.getMessage().contains("LDAP response read timed out")) {
+                            LOGGER.log(Level.WARNING, "LDAP response read time out. AD will fall back to recursive lookup", e);
+                        } else {
+                            throw e;
+                        }
+                    }
                     if (!found || duration >= 10) {
```
Comment (jtnord) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java]: Not this PR - but this logic smells wrong to me.  we could have taken longer than 10s to get all the results - but that may not be due to a read timeout.

Comment (fbelzunc) [inline src/main/java/hudson/plugins/active_directory/ActiveDirectoryUnixAuthenticationProvider.java]: I agree - I will try to fix this very soon. The logic seems to be strange for me as well...

**Pull request merged**